### PR TITLE
Fix concurrency issues with MessageDigest

### PR DESCRIPTION
**Pull request by Warchant**

### CODE REVIEW

File: src/main/java/jp/co/soramitsu/sora/sdk/crypto/json/SelectiveDisclosureFactory.java [hunk 1]
```diff
@@ -22,28 +21,54 @@
...
+    this.saltifier = saltifier;
+    return this;
+  }
+
+  public SelectiveDisclosureFactory withHashAlgorithm(DigestTypeEnum hashAlgorithm) {
+    this.hashAlgorithm = hashAlgorithm;
+    return this;
+  }
+
+  public SelectiveDisclosureFactory withSecurityProvidr(SecurityProvider provider) {
```
Comment (rogachevsn55) [inline src/main/java/jp/co/soramitsu/sora/sdk/crypto/json/SelectiveDisclosureFactory.java]: typo

**Comment by Warchant** [comment]
@rogachevsn55 all classes that have MessageDigest as a member are not thread safe.
Sd factory is, because hashifier (and therefore message digest) is created on each call to sd factory

**Comment by rogachevsn55** [comment]
> @rogachevsn55 all classes that have MessageDigest as a member are not thread safe.
> Sd factory is, because hashifier (and therefore message digest) is created on each call to sd factory

In fact, all of the classes that are marked as such are not ThreadSafe, so I suggest to remove annotation ThreadSafe for them or make code truly thread safe

**Comment by tdorzhi** [comment]
I think what we'd really benefit from is a class producing MessageDigest, probably `Supplier<SHA3.Digest256>`

**Comment by Warchant** [comment]
Well, for this reason I have `SecurityProvider`.
```Java
val provider = new SecurityProvider();
provider.getMessageDigest(/* enum */ SHA3_256); // constructs new MessageDigest of requested type
```

**Pull request merged**